session-replay/Sources/DatadogSessionReplay/Recorder/SnapshotBuilder/NodeRecorders/NodeRecorder.swift

```diff
@@ -0,0 +1,62 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import UIKit
+
+/// A type recording semantics of a given `UIView` or its subclasses. Different implementations of `NodeRecorder` should
+/// recognise specialised subclasses of `UIView` and record their semantics accordingly.
+///
+/// **Note:** The `NodeRecorder` is used on the main thread by `Recorder`.
+internal protocol NodeRecorder {
+    /// The `UIView` subclass this recorder is specialised for.
+    associatedtype View: UIView
+
+    /// Finds the semantic of given`view`.
+    /// - Parameters:
+    ///   - view: the view to determine semantics for
+    ///   - attributes: attributes of this view inferred from its base `UIView` interface
+    ///   - context: the context of building current `ViewTreeSnapshot`
+    /// - Returns: the value of `NodeSemantics` or `nil` if the view is a member of view subclass other than the one this recorder is specialised for.
+    func semantics(of view: View, with attributes: ViewAttributes, in context: ViewTreeSnapshotBuilder.Context) -> NodeSemantics?
+}
+
+/// A type producing SR wireframes.
+///
+/// Each type of UI element (e.g.: label, text field, toggle, button) should provide their own implementaion of `NodeWireframesBuilder`.
+///
+/// **Note:** The `NodeWireframesBuilder` is used on background thread by `Processor`.
+internal protocol NodeWireframesBuilder {
+    /// Creates wireframes that are later uploaded to SR backend.
+    /// - Parameter builder: the generic builder for constructing SR data models.
+    /// - Returns: one or more wireframes that describe a node in SR.
+    func buildWireframes(with builder: WireframesBuilder) -> [SRWireframe]
+}
+
+// MARK: - Type Erasure
+
+/// Type erasure for `NodeRecorder`.
+internal struct AnyNodeRecorder<View> {
+    private let erasedMethod: (View, ViewAttributes, ViewTreeSnapshotBuilder.Context) -> NodeSemantics?
+
+    init<NR: NodeRecorder>(_ concreteRecorder: NR) {
+        self.erasedMethod = { view, node, context in
+            guard let concreteView = view as? NR.View else {
+                // Means that the `UIView` instance passed to erased `NodeRecorder` was not of the type
+                // that this recorder supports. In that case, semantics cannot be determined.
+                return nil
+            }
+            return concreteRecorder.semantics(of: concreteView, with: node, in: context)
+        }
+    }
```

Comment:
Yeah, the main purpose of erasure was to avoid repeating `guard let` in all implementations. But your point is correct - all views are instance of `UIView` and derived type, so it could be avoided. I'm OK to remove erasures for the sake of simplicity 👍 - less constraints could be even better as I can't fully foresee what other challenges are there in UIKit / SwiftUI hierarchies.

Reply:
I went with `guard let ... else { nil }` way - it's simpler and not that tedious anyway 👍.